cmd/crypto/kms.go

```diff
@@ -0,0 +1,139 @@
+// Minio Cloud Storage, (C) 2015, 2016, 2017, 2018 Minio, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crypto
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/minio/minio/cmd/logger"
+	sha256 "github.com/minio/sha256-simd"
+	"github.com/minio/sio"
+)
+
+// Context is a list of key-value pairs cryptographically
+// associated with a certain object.
+type Context map[string]string
+
+// WriteTo writes the context in a canonical from to w.
+// It returns the number of bytes and the first error
+// encounter during writing to w, if any.
+//
+// WriteTo sorts the context keys and writes the sorted
+// key-value pairs as canonical JSON object to w.
```

Comment:
Why sorting is so important here?  Without sorting requirement, this function is not required.

Reply:
As the comment documents it writes the JSON in a "canonical" form. This is required because
the context is cryptographically authenticated / bound to an object. If the context representation (the way it's processed) is not deterministic, verifying integrity would fail.
For example:
Go's `map` guarantees no ordering -> If you hash the key-value pairs twice without any per-processing
/ normalization, the hash values may differ because of ordering. Notice that JSON marshaling / unmarhsaling is not deterministic.

A normalized JSON is not directly required but the advantage is that you can simply write to a local/remote KMS implementation. See [context.WriteTo(mac)](https://github.com/minio/minio/pull/6161/files/d30926a423ecc58dbe1c309ad542ba5a767042e8#diff-4d5e8804f48a7eaf8c84a6e56e59c5b4R136). This avoids some boilerplate serializing/de-serializing code.

It's obvious that such a function wouldn't be required for simple JSON marshaling - (`json.Marshal` / `json.Unmarshal`)

Reply:
I don't know whether I understand it completely here.

It seems `Context` needs to be stored at KMS in a sorted order.

Reply:
Not directly. The KMS uses some crypto (at the moment a crypto, hash function) to process the context. Now if the context would be `{"bucket": "object", "time": "2002-10-02T10:00:00-05:00"}` than hashing the key-value pairs (or the JSON which is just the key-value pairs + `{`, `}`) is not deterministic.
It makes a difference whether you hash the strings: 1. "bucket" 2. "object" 3. "time", ...
or 1. "time", ... 3. "bucket" 4. "object". The hash over the entire context is different in both cases.
It's like `range` over a map. Go doesn't guarantee any ordering but if you want to process data in a order-dependent way you have to somehow canonicalize your processing - For example sorting.